### PR TITLE
Frame: Introduce `TryConvert` and `MaybeConvert` family of traits

### PR DESCRIPTION
**Pull request by gavofyork**
These traits are helpful for cleaner code downstream in Polkadot.

### CODE REVIEW

File: primitives/runtime/src/traits.rs [hunk 1]
```diff
@@ -530,18 +530,68 @@ morph_types! {
...
+}
+
 impl<A, B: Default> Convert<A, B> for () {
 	fn convert(_: A) -> B {
 		Default::default()
 	}
 }
 
+#[impl_trait_for_tuples::impl_for_tuples(30)]
+impl<A, B> TryConvert<A, B> for Tuple {
```
Comment (ggwpez) [inline primitives/runtime/src/traits.rs:570]: I have a hard time imagining how usage of this looks like.  
Could you maybe add a minimal test of `(u8, u8) -> (u16, u16)` or similar?

**Review by ggwpez — APPROVED** [review]

File: primitives/runtime/src/traits.rs (hunk 1 above)
Comment (bkchr) [reply primitives/runtime/src/traits.rs:570]: The idea is `TryConvert<u8, u16> for (TypeX, TypeY)`. This isn't about making `A` and `B` tuples (which is also valid, even without this impl).

File: primitives/runtime/src/traits.rs [hunk 2]
```diff
@@ -530,18 +530,68 @@ morph_types! {
...
 	fn convert(a: A) -> B;
 }
 
+/// Fallible conversion trait returning a Result. Generic over both source and destination types.
+///
+/// In case an `Error` is returned, it must always be the operand which is returned.
+pub trait TryConvert<A, B> {
+	/// Attempt the conversion and return the input value in the error case.
+	fn try_convert(a: A) -> Result<B, A>;
+}
```
Comment (bkchr) [inline primitives/runtime/src/traits.rs:545]:
```suggestion
}

impl<A, B, T: Convert<A, B>> TryConvert<A, B> for T {
    fn try_convert(a: A) -> Result<B, A> {
        Ok(T::convert(a))
    }
}
```

File: primitives/runtime/src/traits.rs [hunk 3]
```diff
@@ -530,18 +530,68 @@ morph_types! {
...
-/// Extensible conversion trait. Generic over both source and destination types.
+/// Infallible conversion trait. Generic over both source and destination types.
 pub trait Convert<A, B> {
 	/// Make conversion.
 	fn convert(a: A) -> B;
 }
 
+/// Fallible conversion trait returning a Result. Generic over both source and destination types.
+///
+/// In case an `Error` is returned, it must always be the operand which is returned.
```
Comment (bkchr) [inline primitives/runtime/src/traits.rs:541]:
```suggestion
```

Already documented above the function?

File: primitives/runtime/src/traits.rs [hunk 4]
```diff
@@ -530,18 +530,68 @@ morph_types! {
...
+///
+/// This specifically reverses the conversion.
+pub trait ConvertBack<A, B>: Convert<A, B> {
+	/// Make conversion back.
+	fn convert_back(b: B) -> A;
+}
+
+/// Fallible conversion trait returning a Result. Generic over both source and destination types.
+///
+/// In case an `Error` is returned, it must always be the operand which is returned.
```
Comment (bkchr) [inline primitives/runtime/src/traits.rs:557]:
```suggestion
```

Same

**Review by bkchr — APPROVED** [review]

File: primitives/runtime/src/traits.rs (hunk 2 above)
Comment (ggwpez) [reply primitives/runtime/src/traits.rs:545]: I also tried this locally, and even after deleting the other implementations, it errors in the tuple impl:

```pre
 error[E0119]: conflicting implementations of trait `TryConvert<_, _>` for type `(_, _)`
     --> work/substrate/primitives/runtime/src/traits.rs:570:1
      |
  570 | / impl<A, B> TryConvert<A, B> for Tuple {
  571 | |     fn try_convert(a: A) -> Result<B, A> {
  572 | |         for_tuples!( #(
  573 | |             let a = match Tuple::try_convert(a) {
  ...   |
  579 | |     }
  580 | | }
      | |_^ conflicting implementation for `(_, _)`
  ...
  648 |   impl<A, B, T: Convert<A, B>> TryConvert<A, B> for T {
      |   --------------------------------------------------- first implementation here
      |
      = note: downstream crates may implement trait `traits::Convert<_, _>` for type `(_, _)`

File: primitives/runtime/src/traits.rs [hunk 5]
```diff
@@ -530,18 +530,64 @@ morph_types! {
...
+
+/// Reversing infallible conversion trait. Generic over both source and destination types.
+///
+/// This specifically reverses the conversion.
+pub trait ConvertBack<A, B>: Convert<A, B> {
+	/// Make conversion back.
+	fn convert_back(b: B) -> A;
+}
+
+/// Fallible conversion trait returning a Result. Generic over both source and destination types.
```
Comment (gavofyork) [inline primitives/runtime/src/traits.rs:553]:
```suggestion
/// Fallible conversion trait returning a [Result]. Generic over both source and destination types.
```

**Pull request merged**